### PR TITLE
Kubevirt network connectivity tests

### PR DESCRIPTION
**Pull request by orenc1**
This PR is a continuation of PR #27308 by @davidvossel 

Introducing 6 test cases to cover pods ability to communicate each other in several scenarios:
* Direct connection between pods on the guest cluster, via Pod Network and Host Network
* Connection between pods in guest cluster to a NodePort service configured on the guest cluster via Pod Network and Host Network.
* Connection between pods in infra cluster to a NodePort service configured on the guest cluster
* ~~Connection between pods in guest cluster to a LoadBalancer service configured on the guest cluster~~
* ~~Connection between pods in infra cluster to a LoadBalancer service configured on the guest cluster~~

### CODE REVIEW

File: test/extended/networking/services.go
Comment (davidvossel) [inline test/extended/networking/services.go:310]: If the LB tests don't pass right now due to issues with cloud-provider-kubevirt, maybe we can just reduce the scope of this to exercising NodePort for now.

File: test/extended/networking/services.go
Comment (davidvossel) [inline test/extended/networking/services.go:295]: does the generic `should allow connections to another pod on a different node via a service IP`  test already cover this?

File: test/extended/networking/services.go
Comment (davidvossel) [inline test/extended/networking/services.go:289]: can we verify this NodePort functionality works when the client is in the pod network as well?

File: test/extended/networking/services.go
Comment (orenc1) [reply test/extended/networking/services.go:295]: Not exactly. In this test case we're using the target pod's internal IP, without exposing it as a service in the cluster.
I agree it should behave the same, since the ServiceIP is being translated to the Pod IP.
If using the direct IP of the target pod doesn't add any value, we can drop it. What do you think?

Comment (davidvossel) [reply test/extended/networking/services.go:295]: ah, alrighty, yeah lets leave the test then

File: test/extended/networking/services.go
Comment (orenc1) [reply test/extended/networking/services.go]: done.

File: test/extended/networking/services.go
Comment (orenc1) [reply test/extended/networking/services.go]: done, removed both LB test cases for now.

**Review by davidvossel — APPROVED** [review]
/lgtm

**Comment by davidvossel** [comment]
@orenc1 needs a rebase. We'll need an approve as well, which i can't provide for this repo.

**Comment by orenc1** [comment]
rebased.
@dcbw / @simonpasquier , could you please review?
The rationale is the expand the network test suites for kubevirt-based hypershift guest OpenShift clusters.
Thanks

**Comment by davidvossel** [comment]
/cc @bparees @mfojtik @knobunc @ironcladlou @squeed @dcbw @danwinship @adambkaplan

Sorry for casting such a wide net here, but we need feedback. @orenc1 has already rebased this PR a few times and I don't want him investing any more time on this if this PR will not be accepted or reviewed.

The Hypershift/KubeVirt platform has some unique test cases that need to be exercised that involve ensuring nested OVNKubernetes works as expected. We have more test cases planned, but we started with this subset of tests that exercise the nodePort connectivity across guest and infra nodes. As a follow up there will be test cases that exercise LB functionality and perhaps network connectivity during and after VM live migration.

Here's the questions we need answered.
1. Is this the right place to introduce these types of hypershift/kubevirt specific tests?
2. what can we do to increase the velocity of our development here with getting similar PRs merged? For example, can we own a directory and do whatever we need to do within it that's specific to kubevirt? or become approvers within our domain (kubevirt)?

Thanks
- David

**Comment by bparees** [comment]
> Is this the right place to introduce these types of hypershift/kubevirt specific tests?

i'd defer to the networking team on that.  I think you want @trozet at this point.

> what can we do to increase the velocity of our development here with getting similar PRs merged? For example, can we own a directory and do whatever we need to do within it that's specific to kubevirt? or become approvers within our domain (kubevirt)?

either ask the networking team to add you as approvers on their tests or extract your tests to their own dir with your own OWNERS file (preferred, but i don't know how easily decoupleable your tests are)

**Comment by bparees** [comment]
also something you can do right now is make sure your team is represented in this files:
https://github.com/openshift/origin/blob/403f0a4fe27e04c4ebbbd3b06fb9ba8639f7ec1c/OWNERS_ALIASES#L14

so that at least you've got coverage for the generated files via:
https://github.com/openshift/origin/blob/403f0a4fe27e04c4ebbbd3b06fb9ba8639f7ec1c/test/extended/util/annotate/generated/OWNERS#L4

**Comment by orenc1** [comment]
Hi @bparees , @davidvossel ,
I've refactored the PR to decouple kubevirt platform related test to another package.
To avoid code duplication, I exported some functions from the `networking` package.
Could you please review?
Thanks

File: test/extended/kubevirt/OWNERS
Comment (bparees) [inline test/extended/kubevirt/OWNERS:7]: i don't need to be here. i'm a top level approver anyway and i have no specific reason to be an approver for this dir.

Comment (orenc1) [reply test/extended/kubevirt/OWNERS]: ok, done.

**Review by davidvossel — COMMENTED** [review]
/lgtm

While we've drastically reduced the amount of code in the standard networking tests, we still need someone to approve the few changes that have been made to the networking utilities to support our new kubevirt related tests.

File: test/extended/kubevirt/services.go
Comment (danwinship) [inline test/extended/kubevirt/services.go:27]: (nitpick: existing style does not start or end blocks with blank lines)

File: test/extended/kubevirt/services.go
Comment (danwinship) [inline test/extended/kubevirt/services.go:37]: `Expect(err).NotTo(HaveOccurred())`

you can do

```
err := functionThatReturnsAnError()
Expect(err).NotTo(HaveOccurred())
```

or

```
Expect(functionThatReturnsAnError()).To(Succeed())
```

but doing `Expect(err).To(Succeed())` makes it confusing to read. (What does it mean for an error to succeed?)

(This comment applies in a bunch of places.)

File: test/extended/kubevirt/util.go
Comment (danwinship) [inline test/extended/kubevirt/util.go:44]: you're not actually using all of these constants that you copied over

File: test/extended/kubevirt/util.go
Comment (danwinship) [inline test/extended/kubevirt/util.go:57]: FTR there are almost certainly helper methods in `k8s.io/kubernetes/test/e2e/framework/{pod,network,service}` that you could be using instead of defining `launchWebserverNodePortService`, `launchWebserverLoadBalancerService`, `checkConnectivityToHostWithCLI` and calling `networking.CreatePodForService` and `networking.CheckConnectivityToHost`

(There's a jira issue open about porting the existing old networking tests away from the legacy helper methods. It would be good to at least not add _more_ tests using those methods (and more non-standard methods!).)

File: test/extended/kubevirt/util.go
Comment (danwinship) [inline test/extended/kubevirt/util.go:157]: `e2enode.GetRandomReadySchedulableNode()`
(which is actually better than this function because it returns a random node rather than always returning the first node, which makes for better e2e test parallelizability)

File: test/extended/kubevirt/util.go
Comment (danwinship) [inline test/extended/kubevirt/util.go:195]: This is returning the name of the first node that _isn't_ `serverInfraNode`, which... is not what the function name sounds like it is supposed to be doing?

File: test/extended/kubevirt/util.go
Comment (danwinship) [inline test/extended/kubevirt/util.go:260]: Ugh, `findAppropriateNodes` is a disaster; it tries to strip master and infra nodes from the list, except that it uses the 3.x node label values rather than the 4.x ones so it totally fails. :joy: And so of course, then the tests that use `findAppropriateNodes` end up having to use `makeNamespaceScheduleToAllNodes` because they accidentally schedule random pods to the masters...

If you use `e2enode.GetBoundedReadySchedulableNodes(serverFramework.ClientSet, 2)`, you'll get back 2 nodes that don't include any tainted nodes (and thus don't include the masters), and then you don't need to call `makeNamespaceSchedulableToAllNodes` either. (You shouldn't need to call that in any other tests either.)

File: test/extended/kubevirt/util.go
Comment (danwinship) [inline test/extended/kubevirt/util.go:237]: This is probably consistent with the other networking tests, but it's unnecessary; the e2e framework deletes everything in the test case's namespace for you. (I'm not sure if this is a "new" behavior that was added after the original network tests were written, or if the network tests were always wrong.)

File: test/extended/kubevirt/util.go
Comment (danwinship) [inline test/extended/kubevirt/util.go:245]: If you're parsing a URL, use a URL-parsing function. (And if not, at least use `net.SplitHost` rather than `strings.Split`)

File: test/extended/networking/util.go
Comment (danwinship) [inline test/extended/networking/util.go:98]: This is no longer needed in `networking/util.go` since you copied the functions out into `kubevirt/util.go` (except that you shouldn't use them there either :slightly_smiling_face: )

**Review by danwinship — COMMENTED** [review]
It would be better to squash the commits together so that you're just creating the new tests in the right place rather than creating them in the "wrong" place in one commit and then immediately moving them.

**Comment by danwinship** [comment]
We really need to put a warning on the networking tests that they are all terrible and no one should use them as examples of how to write test cases :slightly_frowning_face:

File: test/extended/kubevirt/util.go
Comment (orenc1) [reply test/extended/kubevirt/util.go:44]: yes, right.
eventually I used the consts directly from the networking package.

File: test/extended/kubevirt/util.go
Comment (orenc1) [reply test/extended/kubevirt/util.go:195]: We want to get a schedulable infra node for the client pod, that is different than the node the server pod is running on.
note the function name implies we want to get a `SchedulableInfraNode`, and we're checking that node is not the infra node that running the **server** [pod]

File: test/extended/kubevirt/util.go
Comment (orenc1) [reply test/extended/kubevirt/util.go:237]: it was added for precaution.
if you say the framework is already doing the cleanup, i'll remove those.

File: test/extended/kubevirt/util.go
Comment (orenc1) [reply test/extended/kubevirt/util.go:57]: i tried to find at `k8s.io/kubernetes/test/e2e/framework/*` functions that enabled us to create services using NodePort, but there seems to be anything close enough for our use case.
but i can dig more into it in a followup PR, since anyway we'll need to add the guest/infra LoadBalancer service connectivity tests. we'll try to utilize the helper methods from there.

**Comment by orenc1** [comment]
Hi @danwinship ,
Thank you very much for your review and insightful comments!
I think I was able to address them, I squashed everything to one commit and reduce the changes at `test/extended/networking` to the bare minimum.
could you please re-review?

**Comment by orenc1** [comment]
Hi @davidvossel , @bparees ,
I've completely decoupled ourselves from the `networking` package, not changes are done there at all.
I had to copy over few functions (`createPodForService`, `platformType` and `checkConnectivityToHost` with slight modification).

**Review by davidvossel — COMMENTED** [review]
/lgtm

**Pull request merged**
